mikrotik/resource_generic_crud_operations.go

```diff
@@ -0,0 +1,127 @@
+package mikrotik
+
+import (
+	"context"
+
+	"github.com/ddelnano/terraform-provider-mikrotik/client"
+	"github.com/ddelnano/terraform-provider-mikrotik/mikrotik/internal/utils"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+type (
+	CreateFunc func(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse)
+	ReadFunc   func(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse)
+	UpdateFunc func(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse)
+	DeleteFunc func(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse)
+)
+
+// GenericCreateResource creates the resource and sets the initial Terraform state.
+//
+// terraformModel and mikrotikModel must be passed as pointers
+func GenericCreateResource(terraformModel interface{}, mikrotikModel client.Resource, client *client.Mikrotik) CreateFunc {
+	return func(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+
+		diags := req.Plan.Get(ctx, terraformModel)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		if err := utils.CopyStruct(terraformModel, mikrotikModel); err != nil {
+			resp.Diagnostics.AddError("Cannot copy model: Terraform -> MikroTik", err.Error())
+			return
+		}
+
+		created, err := client.Add(mikrotikModel)
+		if err != nil {
+			resp.Diagnostics.AddError("Creation failed", err.Error())
+			return
+		}
+
+		if err := utils.CopyStruct(created, terraformModel); err != nil {
+			resp.Diagnostics.AddError("Cannot copy model: MikroTik -> Terraform", err.Error())
+			return
+		}
+
+		resp.Diagnostics.Append(resp.State.Set(ctx, terraformModel)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+}
+
+// GenericReadResource refreshes the Terraform state with the latest data.
+func GenericReadResource(terraformModel interface{}, mikrotikModel client.Resource, client *client.Mikrotik) ReadFunc {
+	return func(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+		resp.Diagnostics.Append(req.State.Get(ctx, terraformModel)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		if err := utils.CopyStruct(terraformModel, mikrotikModel); err != nil {
+			resp.Diagnostics.AddError("Cannot copy model: Terraform -> MikroTik", err.Error())
+			return
+		}
```

Comment:
I find the flip flopped `utils.CopyStruct` to be slightly confusing to follow (doing it to the mikrotik model first, followed by the terraform one). However since we are implementing this logic in a single place, it should be easy to ensure they don't get misused (accidentally switching src and dest).

I'm fine to leave this code as is, but it might be easier to follow the implmenentation of the `CopyStruct` func and this logic if they were two separate functions.

Reply:
@ddelnano I thought about that as well, but decided to leave it as-is.
Since it is not obvious during reviewing, I think it's worth to make it more type-safe.
Let me know what you think about the latest commit.

Reply:
Latest commit lgtm.